### DISCUSSION TITLE
Proposal: Clippath stroke handling

### DISCUSSION THREAD

**Discussion by melchiar** [fabricjs/fabric.js#6865] [General]
I'm looking for some feedback on possible adjustments to how Fabric handles clipPath strokes. Currently, clipPath strokes are ignored by the clipped object, which I feel could be improved upon. 

In my own project I've overridden this functionality so that a clipPath's stroke extends the clipping area, which is useful to simulate document bleed on A clipped canvas.  I also plan to add a property to reverse this effect so that a clipPath's stroke shrinks the clipping area rather than expanding it. This will allow for adding an inset border to the canvas without having to do the hard work of recalculating the border's point coordinates according to the correct distance from the edges.

In the images below, the dotted line indicates the path of the clipPath object. 

![invertstroke-mockup](https://user-images.githubusercontent.com/9057412/107836467-14e17d80-6d52-11eb-8d4a-023d83f5c1e7.jpg)

**Questions.** 

1. To my knowledge, this functionality would not be entirely reproducible in svg. While I believe svg clippaths do include strokes in the clipping region, there is no way that I'm aware of to support svg clippaths with inverted strokes. I believe the current `inverted` property is also not supported in SVG, so an `invertedStroke` property would have a similar limitation. Does anyone see this as an issue?
2. Adding this could possibly interfere with any future plans to allow clipPaths to have filled strokes (ie. clipped images with strokes that follow the shape of the clipPath). Not sure if there are any plans to add this in future, but I know I've seen it requested before.

Thanks!

**Comment by asturur** [comment]
In my opinion clipPath are a replacement for clipTo and a compatibility function for SVG.
The right thing to do in order to make your feature work are masks in this PR here:
https://github.com/fabricjs/fabric.js/pull/6310
This works already, works as clipPaths ( you can see from the code that most of the logic is exactly the same) but it does include borders and can also make use of gradients and patterns.

https://github.com/fabricjs/fabric.js/pull/6310/files#diff-de494ae6278212520cae6374e5a8eecaa8d08248ab53f6f30c167949186292caR1243

If you have a clipPath, clipPath still wins.

Mask in few words should be a multiply operation mostly ( i have to double check ) so white will keep things as they are, transparent will make transparent, grey will change colors, black will make things black. other colors will make other effects on the image.

But if you use it with white and transparency is an easy tool to obtain effects you want.

If you can write code to change how clipPath works, you should really look into  updating that PR with latest code and give it a shot.

**Reply by melchiar** [reply]
Good idea, I hadn't considered trying masking. I will give it a go.
